### PR TITLE
Synchronize cpp examples with python

### PR DESCRIPTION
**Pull request by csaba-luxonis**
This PR is WIP, will be updated once it's ready to review.

### CODE REVIEW

**Comment by Luxonis-Brandon** [comment]
Thanks for making these.  These will be great to have!  Will be so nice to have full Python and C++ examples for folks.

**Comment by Luxonis-Brandon** [comment]
Oh and once these are done @csaba-luxonis - we should do the switcher on our docs so folks can switch between the Python code sample and the C++ code sample on the same page.

**Comment by csaba-luxonis** [comment]
Yeah, that would be amazing.

**Comment by themarpe** [comment]
I think we should do this in other direction:
Remove example numbers from python examples and leave these here as is.

Documentation should mark examples (code samples) with numbers but just as an easier overview

**Comment by Luxonis-Brandon** [comment]
Oh, yes.  Sorry I didn't meant to say that we should have the numbers on these.  I agree we should remove the numbers from the Python examples.  And just have the documentation designate the order of the code samples.

File: examples/src/04_rgb_encoding.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
+
+#include <csignal>
+#include <cstdio>
+#include <iostream>
+
+#include "utility.hpp"
```
Comment (themarpe) [inline examples/src/04_rgb_encoding.cpp:6]: Unneeded

File: examples/src/04_rgb_encoding.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+// Inludes common necessary includes for development using depthai library
+#include "depthai/depthai.hpp"
+
+    // Keyboard interrupt (Ctrl + C) detected
```
Comment (themarpe) [inline examples/src/04_rgb_encoding.cpp:11]: Remove leading whitespace

File: examples/src/04_rgb_encoding.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    // Output queue will be used to get the encoded data from the output defined above
+    auto q = d.getOutputQueue("h265", 30, true);
+
+    // The .h265 file is a raw stream file (not playable yet)
+    auto videoFile = std::fstream(h265Path, std::ios::out | std::ios::binary);
+    std::cout << "Press Ctrl+C to stop encoding..." << std::endl;
+
+    while(alive) {
+        auto h264Packet = q->get<dai::ImgFrame>();
+        videoFile.write((char*)h264Packet->getData().data(), h264Packet->getData().size());
```
Comment (themarpe) [inline examples/src/04_rgb_encoding.cpp:60]:
```
videoFile.write(static_cast<char*>(h264Packet->getData().data()), h264Packet->getData().size());
```

File: examples/src/04_rgb_encoding.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    }
+    videoFile.close();
```
Comment (themarpe) [inline examples/src/04_rgb_encoding.cpp:62]: Unneeded - RAII should take care of it

File: examples/src/04_rgb_encoding.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    // Pipeline is defined, now we can connect to the device
+    dai::Device d(p);
+    // Start pipeline
+    d.startPipeline();
+
+    // Output queue will be used to get the encoded data from the output defined above
+    auto q = d.getOutputQueue("h265", 30, true);
+
+    // The .h265 file is a raw stream file (not playable yet)
+    auto videoFile = std::fstream(h265Path, std::ios::out | std::ios::binary);
```
Comment (themarpe) [inline examples/src/04_rgb_encoding.cpp:55]:
```
auto videoFile = std::ofstream(h265Path, std::ios::binary);
```

File: examples/src/04_rgb_encoding.cpp [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+#include <cstdio>
+#include <iostream>
+
+#include "utility.hpp"
+
+// Inludes common necessary includes for development using depthai library
+#include "depthai/depthai.hpp"
+
+    // Keyboard interrupt (Ctrl + C) detected
+static bool alive = true;
```
Comment (themarpe) [inline examples/src/04_rgb_encoding.cpp:12]:
```
static std::atomic<bool> alive{true};
```

File: examples/src/04_rgb_encoding.cpp [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+static void sigintHandler(int signum) {
+    alive = false;
+}
+
+int main(int argc, char** argv) {
+    using namespace std;
+    using namespace std::chrono;
+    signal(SIGINT, &sigintHandler);
```
Comment (themarpe) [inline examples/src/04_rgb_encoding.cpp:20]: Maybe a small comment on this + remove the using std and std::chrono, we don't use it here actually
```
std::signal(SIGINT, &sigintHandler);
```

File: examples/src/04_rgb_encoding.cpp [hunk 8]
```diff
@@ -0,0 +1,68 @@
+
+#include <csignal>
+#include <cstdio>
+#include <iostream>
+
+#include "utility.hpp"
+
+// Inludes common necessary includes for development using depthai library
```
Comment (themarpe) [inline examples/src/04_rgb_encoding.cpp:8]: Typo: `Inludes` -> `Includes`

File: examples/src/04_rgb_encoding.cpp [hunk 9]
```diff
@@ -0,0 +1,68 @@
+
```
Comment (themarpe) [inline examples/src/04_rgb_encoding.cpp:1]: Extra line

**Review by themarpe — CHANGES_REQUESTED** [review]

File: examples/src/04_rgb_encoding.cpp (hunk 3 above)
Comment (csaba-luxonis) [reply examples/src/04_rgb_encoding.cpp:53]: static_cast from 'unsigned char *' to 'char *' is not allowed

Comment (themarpe) [reply examples/src/04_rgb_encoding.cpp]: Good catch, `reinterpret_cast<char*>` should do

I guess we can leave C-style cast in this case

**Comment by themarpe** [comment]
Also one thing I was thinking, if we allow passing the output file as parameter (as well as the default), to be able to set it to `/dev/null` / `nul` for testing case. (Same applies to Python side)

**Review by themarpe — CHANGES_REQUESTED** [review]
Throughout examples: use ImgFrame::Type alias instead of Raw*
```
manip->initialConfig.setFrameType(dai::ImgFrame::Type::BGR888p);
```

**Review by themarpe — DISMISSED** [review]
LGTM apart from:

Style - either same line or with curly braces (in a couple of cases)
```
if(syncNN) {
    ...
} else {
    ...
}
```

**Review by szabi-luxonis — APPROVED** [review]
LGTM

**Pull request merged**